include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -3509,6 +3511,317 @@ namespace internal
+
+      // h1 elements have hirarchic numbering, l2 elements have lexiographic
+      // numbering therefore, we need to convert the dof indices
+      const auto l2_to_h1 =
+        FETools::lexicographic_to_hierarchic_numbering<dim>(degree);
+
+      using DoFCellIterator =
+        typename DoFHandler<dim, spacedim>::active_cell_iterator;
+
+      std::vector<std::pair<unsigned int, types::global_dof_index>>
+        sp_to_l2_dofs;
+      sp_to_l2_dofs.reserve(dof_handler_l2.n_locally_owned_dofs());
+
+      // fill sp_to_l2_dofs
+      std::vector<types::global_dof_index> dof_indices_h1(dofs_per_cell);
+      std::vector<types::global_dof_index> dof_indices_l2(dofs_per_cell);
+      for (const auto &cell : tria.active_cell_iterators())
+        {
+          if (cell->is_locally_owned() || cell->is_ghost())
+            {
+              DoFCellIterator cell_h1 = {&tria,
+                                         cell->level(),
+                                         cell->index(),
+                                         &dof_handler_h1};
+              DoFCellIterator cell_l2 = {&tria,
+                                         cell->level(),
+                                         cell->index(),
+                                         &dof_handler_l2};
+              cell_h1->get_dof_indices(dof_indices_h1);
+              cell_l2->get_dof_indices(dof_indices_l2);
+
+              // collect unconstrained dofs for support point
+              for (unsigned int i = 0; i < dof_indices_h1.size(); ++i)
+                if (partitioner_h1.in_local_range(dof_indices_h1[i]))
+                  {
+                    const auto global_l2 = dof_indices_l2[l2_to_h1[i]];
+                    if (!constraint.is_constrained(global_l2))
+                      {
+                        const auto local_h1 =
+                          partitioner_h1.global_to_local(dof_indices_h1[i]);
+                        sp_to_l2_dofs.emplace_back(local_h1, global_l2);
+                      }
+                  }
+            }
+        }
+
+      // sort for support points
+      std::sort(sp_to_l2_dofs.begin(),
+                sp_to_l2_dofs.end(),
+                [](const auto &a, const auto &b) { return a.first < b.first; });
+
+
+      // convert to CRS format
+      auto                                 it = sp_to_l2_dofs.begin();
+      std::vector<types::global_dof_index> l2_dofs;
+      l2_dofs.reserve(sp_to_l2_dofs.size());
+      std::vector<unsigned int> l2_dof_ptrs;
+      l2_dof_ptrs.reserve(dof_handler_h1.n_locally_owned_dofs() + 1);
+      l2_dof_ptrs.emplace_back(0);
+      for (unsigned int i = 0; i < dof_handler_h1.n_locally_owned_dofs(); ++i)
+        {
+          const unsigned int index = it->first;
+          while (it != sp_to_l2_dofs.end() && it->first == index)
+            {
+              l2_dofs.emplace_back(it->second);
+              ++it;
+            }
+          l2_dof_ptrs.emplace_back(l2_dofs.size());
+        }
```

Comment:
At the same time, you could determine for each local support point (`it->first`) the actual index (`l2_dof_ptrs.size()`). This mapping could be used set the point vector.

Reply:
I restructured a big portion of code to achieve this :) ([f85dd34](https://github.com/dealii/dealii/pull/15209/commits/f85dd34bc167f7f13689909da7fda3bb07dd2ec4))